src/backend/commands/copy.c

```diff
@@ -4473,6 +4480,68 @@ CopyFrom(CopyState cstate)
 	return processed;
 }
 
+/*
+ * A subroutine to update the insert counts in all partitions for AO relations.
+ */
+static void CopyFromUpdateAOInsertCountInPartitions(HTAB *partition_hash, 
+													HTAB *tupcounts_ht,
+													List *ao_segnos)
+{
+	HASH_SEQ_STATUS hash_seq_status;
+	ResultPartHashEntry *entry;
+
+	hash_seq_init(&hash_seq_status, partition_hash);
+	while ((entry = hash_seq_search(&hash_seq_status)) != NULL)
+	{
+		ResultRelInfo *partInfo = &entry->resultRelInfo;
+
+		if (relstorage_is_ao(RelinfoGetStorage(partInfo)))
+		{
+			bool partFound;
+			Oid partRelid = RelationGetRelid(partInfo->ri_RelationDesc);
+			uint64 partTupcount = CopyFromGetTupCount(tupcounts_ht,
+													  partRelid, &partFound);
+
+			if (partFound)
+			{
+				ResultRelInfoSetSegno(partInfo, ao_segnos);
+				CopyFromUpdateAOInsertCount(partInfo, partTupcount);
+			}
+		}
+	}
+}
+
+/*
+ * Fetches the tuple count for a given relation from the hash table.
+ * The `found` parameter is set by `hash_search`, which performs
+ * the lookup in the hash table. Returns the tuple count if found,
+ * otherwise returns 0.
+ */
+static int64 CopyFromGetTupCount(HTAB *ht, Oid relid, bool *found)
+{
+	struct {
+		Oid relid;
+		int64 tupcount;
+	} *ao;
+
+	ao = hash_search(ht, &relid, HASH_FIND, found);
```

Comment:
why not
```suggestion
	PQaoRelTupCount *ao = hash_search(ht, &relid, HASH_FIND, found);
```

Reply:
Ok, now we can remove the CopyFromGetTupCount function because it only misleads the code.